rsconnect/metadata.py

```diff
@@ -420,3 +430,203 @@ def resolve(self, server, app_id, app_mode):
+    CONNECT_ADMIN_REBUILD_DIR or the current working directory is none is supplied.
+
+    A single `rebuild.json` file contains "tracked" content for 1-N connect servers.
+    The structure is as follows:
+    {
+        "<connect server url>": {
+            "rsconnect_rebuild_running": <bool>,
+            "<content guid 1>": {
+                "rsconnect_rebuild_status": <models.RebuildStatus>,
+                ..., // various content metadata fields returned by the v1/content api
+            },
+            "<content guid 2>": {
+                ...,
+            }
+        },
+        "https://another-connect-server:443": {
+            ...,
+        },
+        ...
+    }
+    """
+
+    _REBUILD_ABORTED = False
+
+    def __init__(self, base_dir=os.getenv("CONNECT_ADMIN_REBUILD_DIR", DEFAULT_REBUILD_DIR)):
+        self._base_dir = base_dir
+        super(ContentRebuildStore, self).__init__(join(base_dir, "rebuild.json"), chmod=True)
+
+    def aborted(self):
+        return ContentRebuildStore._REBUILD_ABORTED
+
+    def get_rebuild_logs_dir(self, server, guid):
+        return join(self._base_dir, "logs", _normalize_server_url(server.url), guid)
+
+    def ensure_logs_dir(self, server, guid):
+        log_dir = self.get_rebuild_logs_dir(server, guid)
+        os.makedirs(log_dir, exist_ok=True)
+        if self._chmod:
+            os.chmod(log_dir, 0o700)
+
+    def get_rebuild_log(self, server, guid, task_id=None):
+        """
+        Returns the path to the rebuild log file. This method does not check
+        whether the file exists if a task_id is provided.
+        If task_id is not provided, we will return the latest log
+        using the last modified timestamp on the filesystem.
+        If no log files are found, returns None
+        """
+        log_dir = self.get_rebuild_logs_dir(server, guid)
+        if task_id:
+            return join(log_dir, "%s.log" % task_id)
+        else:
+            try:
+                log_files = glob.glob(join(log_dir, "*.log"))
+                latest = max(log_files, key=os.path.getctime)
+            except ValueError:
+                latest = None
+            return latest
+
+    def get_rebuild_history(self, server, guid):
+        """
+        Returns the rebuild history for a given content guid.
+        """
+        log_dir = self.get_rebuild_logs_dir(server, guid)
+        log_files = glob.glob(join(log_dir, '*.log'))
+        history = []
+        for f in log_files:
+            task_id = basename(f).split('.log')[0]
+            t = datetime.fromtimestamp(os.path.getctime(f), tz=timezone.utc).strftime('%Y-%m-%d %H:%M:%S%z')
+            history.append({'time': t, 'task_id': task_id})
+        history.sort(key=lambda x: x['time'])
+        return history
+
+    def get_rebuild_running(self, server):
+        return self._data.get(server.url, {}).get('rsconnect_rebuild_running')
+
+    def set_rebuild_running(self, server, is_running):
+        self._lock.acquire()
```

Comment:
The lock/unlock sequences look correct. Since they are held locally, you have the option to simplify them using a `with` clause as described here: https://docs.python.org/3/library/threading.html#using-locks-conditions-and-semaphores-in-the-with-statement

```
with self._lock:
    self._data[server.url]['rsconnect_rebuild_running'] = is_running
    self.save()
```